src/nginx.c

```diff
@@ -73,108 +71,243 @@ static size_t nginx_curl_callback(void *buf, size_t size, size_t nmemb,
-  else if (strcasecmp(key, "verifyhost") == 0)
-    return config_set(&verify_host, value);
-  else if (strcasecmp(key, "cacert") == 0)
-    return config_set(&cacert, value);
-  else if (strcasecmp(key, "timeout") == 0)
-    return config_set(&timeout, value);
-  else if (strcasecmp(key, "socket") == 0)
-    return config_set(&sock, value);
-  else
+static int nginx_add_daemon_config(oconfig_item_t *ci) {
+  int ret;
+
+  /* Get instance name first - we need it for building metrics prefix as well */
+
+  if ((ci->values_num != 1) || (ci->values[0].type != OCONFIG_TYPE_STRING)) {
+    WARNING("nginx plugin: `InstanceName' block must have only name as an "
+            "argument.");
     return -1;
-} /* int config */
+  }
+
+  ret = nginx_handle_str(ci, ngx_inst[instance_num_counter].name,
+                         DATA_MAX_NAME_LEN);
+  if (ret) {
+    return ret;
+  }
+
+  if (ngx_inst[instance_num_counter].name[0] == '\0') {
+    ERROR("nginx plugin: you must configure nginx instance name.\n");
+    return -EINVAL;
+  }
+
+  /*****************************************************************************/
+
+  /* Initialize CURL structure for the conf and set up default options */
 
-static int init(void) {
-  if (curl != NULL)
-    curl_easy_cleanup(curl);
+  if (ngx_inst[instance_num_counter].curl != NULL)
+    curl_easy_cleanup(ngx_inst[instance_num_counter].curl);
 
-  if ((curl = curl_easy_init()) == NULL) {
+  if ((ngx_inst[instance_num_counter].curl = curl_easy_init()) == NULL) {
     ERROR("nginx plugin: curl_easy_init failed.");
     return -1;
   }
 
-  curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
-  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, nginx_curl_callback);
-  curl_easy_setopt(curl, CURLOPT_USERAGENT, COLLECTD_USERAGENT);
-  curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, nginx_curl_error);
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_NOSIGNAL, 1L);
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_WRITEFUNCTION,
+                   nginx_curl_callback);
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_USERAGENT,
+                   COLLECTD_USERAGENT);
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_ERRORBUFFER,
+                   nginx_curl_error);
+
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_FOLLOWLOCATION,
+                   1L);
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_MAXREDIRS, 50L);
+
+  /**********************************************************************/
+
+  /* Local arrays for temporal storing of the parameters available through
+   * plugin options */
+
+  char u[DATA_MAX_NAME_LEN] = {'\0'};
+  char p[DATA_MAX_NAME_LEN] = {'\0'};
+  char url[DATA_MAX_NAME_LEN] = {'\0'};
+  char verify_peer[DATA_MAX_NAME_LEN] = {'\0'};
+  char verify_host[DATA_MAX_NAME_LEN] = {'\0'};
+  char cacert[DATA_MAX_NAME_LEN] = {'\0'};
+  char timeout[DATA_MAX_NAME_LEN] = {'\0'};
+  char socket[DATA_MAX_NAME_LEN] = {'\0'};
```

Comment:
This function is nearly two hundred lines.  This could be good place to split it (at least) into two.

Single char variables should be used only as loop counters (i, j etc), anything else should have more descriptive names (especially in function this large).

`socket` overrides standard C-lib function name.  Maybe "sockname"?

Otherwise this looks pretty OK on quick glance.

Reply:
hank you for review and for your notes @eero-t .
Definitely you are right about nginx_add_daemon_config() function size and not very obvious names for temporary arrays. So in order to solve the second issue instead of using predefined number of static arrays I decided to return back ```static char *config_keys[]``` setting list and use dynamic ```**parsed_values_for_nginx_instance``` string list for parsing configuration file and further curl options settings. Additionally in terms of the first issue curl options were splitted into default ones ( that cannot be changed from configuration file ) and optional ones that are managed with config.